### ISSUE TITLE
[sklearn#34074] KMeans can skip final label reassignment after empty-cluster relocation and return stale labels_/inertia_

### ISSUE THREAD

**Issue by godofecht** [flooooooooooow/flow#355]
**Mirrored from [scikit-learn#34074](https://github.com/scikit-learn/scikit-learn/issues/34074)**

Original title: KMeans can skip final label reassignment after empty-cluster relocation and return stale labels_/inertia_
Original labels: Bug, Needs Decision, module:cluster
Created: 2026-05-21T13:09:49Z

---

> [!WARNING]
> This issue is not yet ready for a PR. If you are interested in contributing to scikit-learn, please have a look at our [contributing guidelines](https://scikit-learn.org/dev/developers/contributing.html), and in particular the sections for [new contributors](https://scikit-learn.org/dev/developers/contributing.html#new-contributors) and the ["Needs triage"](https://scikit-learn.org/dev/developers/contributing.html#issues-tagged-needs-triage) label.

### Describe the bug and give evidence about its user-facing impact

I think there is a convergence bug in `KMeans` when empty-cluster relocation happens during an iteration.

In both `_kmeans_single_lloyd` and `_kmeans_single_elkan`, the main loop checks for strict convergence with:

```python
if np.array_equal(labels, labels_old):
    strict_convergence = True
    break
```

This check happens after the centers have been updated/swapped. If `strict_convergence` is true, the final E-step is skipped. That final E-step is otherwise used to make sure the returned `labels_` match the final `cluster_centers_`.

The problem is that empty-cluster relocation can mutate `centers_new` and `weight_in_clusters` without mutating `labels`. So `labels` can be unchanged from the previous iteration, while the centers have still changed because an empty cluster was repaired. In that case, the algorithm declares “strict convergence” even though the returned centers and labels are inconsistent.

Relevant code paths:

- `_kmeans_single_lloyd` checks `np.array_equal(labels, labels_old)` and skips the final E-step on strict convergence:
  https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/cluster/_kmeans.py
- `_kmeans_single_elkan` has the same strict-convergence/final-E-step structure:
  https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/cluster/_kmeans.py
- `_relocate_empty_clusters_dense` / `_relocate_empty_clusters_sparse` modify `centers_new` and `weight_in_clusters`, using `labels`, but do not update `labels`:
  https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/cluster/_k_means_common.pyx

### Steps/Code to Reproduce

```python
import numpy as np
from scipy import sparse
from sklearn.cluster import KMeans

X = sparse.csr_matrix(np.array([[1.0], [2.0], [3.0]]))
init = np.array([[4.0], [0.0], [1.0]])

for algorithm in ["lloyd", "elkan"]:
    km = KMeans(
        n_clusters=3,
        init=init,
        n_init=1,
        max_iter=10,
        tol=0,
        algorithm=algorithm,
        verbose=1,
    ).fit(X)

    print("\n", algorithm)
    print("centers: ", km.cluster_centers_.ravel())
    print("labels_: ", km.labels_)
    print("predict: ", km.predict(X))
    print("inertia_:", km.inertia_)
    print("score:   ", km.score(X))
```

### Expected Results

After `fit`, the returned `labels_`, `cluster_centers_`, and `inertia_` should be mutually consistent.

In particular, these should pass:

~~~python
np.testing.assert_array_equal(km.labels_, km.predict(X))
np.testin

...(truncated)

---

## Flow context

This issue is mirrored from scikit-learn to explore whether Flow encounters the same problem or can approach it differently. Add analysis below.

**Comment by godofecht** [comment]
Moved to [64](https://github.com/godofecht/flow-scikit/issues/64) in the flow-scikit repository. Further library-specific triage and implementation will happen there.
